flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisProducerTest.java

```diff
@@ -49,61 +71,205 @@ public void testCreateWithNonSerializableDeserializerFails() {
+			new OneInputStreamOperatorTestHarness<>(new StreamSink<>(producer));
+
+		testHarness.open();
+
+		testHarness.processElement(new StreamRecord<>("msg-1"));
+
+		producer.getPendingRecordFutures().get(0).setException(new Exception("artificial async exception"));
+
+		try {
+			testHarness.snapshot(123L, 123L);
+		} catch (Exception e) {
+			// the next invoke should rethrow the async exception
+			Assert.assertTrue(ExceptionUtils.findThrowableWithMessage(e, "artificial async exception").isPresent());
+
+			// test succeeded
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	/**
+	 * Test ensuring that if an async exception is caught for one of the flushed requests on checkpoint,
+	 * it should be rethrown; we set a timeout because the test will not finish if the logic is broken.
+	 *
+	 * <p>Note that this test does not test the snapshot method is blocked correctly when there are pending recorrds.
+	 * The test for that is covered in testAtLeastOnceProducer.
+	 */
+	@SuppressWarnings("ResultOfMethodCallIgnored")
+	@Test
+	public void testAsyncErrorRethrownAfterFlush() throws Throwable {
+		final DummyFlinkKinesisProducer<String> producer = new DummyFlinkKinesisProducer<>(new SimpleStringSchema());
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness =
+			new OneInputStreamOperatorTestHarness<>(new StreamSink<>(producer));
+
+		testHarness.open();
+
+		testHarness.processElement(new StreamRecord<>("msg-1"));
+		testHarness.processElement(new StreamRecord<>("msg-2"));
+		testHarness.processElement(new StreamRecord<>("msg-3"));
+
+		// only let the first record succeed for now
+		UserRecordResult result = mock(UserRecordResult.class);
+		when(result.isSuccessful()).thenReturn(true);
+		producer.getPendingRecordFutures().get(0).set(result);
+
+		CheckedThread snapshotThread = new CheckedThread() {
+			@Override
+			public void go() throws Exception {
+				// this should block at first, since there are still two pending records that needs to be flushed
+				testHarness.snapshot(123L, 123L);
+			}
+		};
+		snapshotThread.start();
+
+		// let the 2nd message fail with an async exception
+		producer.getPendingRecordFutures().get(1).setException(new Exception("artificial async failure for 2nd message"));
+		producer.getPendingRecordFutures().get(2).set(mock(UserRecordResult.class));
+
+		try {
+			snapshotThread.sync();
+		} catch (Exception e) {
+			// the next invoke should rethrow the async exception
+			e.printStackTrace();
+			Assert.assertTrue(ExceptionUtils.findThrowableWithMessage(e, "artificial async failure for 2nd message").isPresent());
+
+			// test succeeded
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	/**
+	 * Test ensuring that the producer is not dropping buffered records;
+	 * we set a timeout because the test will not finish if the logic is broken.
+	 */
+	@SuppressWarnings("unchecked")
+	@Test(timeout = 10000)
+	public void testAtLeastOnceProducer() throws Throwable {
```

Comment:
I am not sure if this test is good enough:

1. It is not testing the code that it should :( it is using overriden version of the `FlinkKinesisProducer` - `DummyFlinkKinesisProducer` can hide bugs in the real implementation.
2. Implementing it as a unit test with mocks, doesn't test for out integration with `Kinesis`. You made some assumption how `at-least-once` should be implemented, you implemented it in production code and here you are repeating the same code using the same assumptions :(

However looking at Kafka tests instability I'm not sure which approach is worse... Unless those are not tests instabilities but bugs in our code, which Kafka's ITCases are triggering from time to time - this mockito based test would not discover such bugs.

Reply:
I can see your point. Regarding your concerns:

For 1.: I think it is still ok, since the `DummyFlinkKinesisProducer` only overrides the `getKinesisProducer` method to implement a mock producer. Also, while the `snapshotState` method is overriden, I'm only overriding it to inject sync-point latches. The flushing behaviour of the `snapshotState`'s implementation should still be guarded by this test.
For 2.: The lack of integration tests with Kinesis has always been an issue. There simply is no simple way to implement IT tests for that.

What do you think?

Reply:
I am still not convinced on the value of such tests (reverse implementing the production code), but I will not press it since:
> There simply is no simple way to implement IT tests for that.